### PR TITLE
full lazy init of HSAContext

### PR DESCRIPTION
**Pull request by jeffdaily**
HSAContext is no longer a module static object to be initialized during C++ static initialization.
HCC_LAZYINIT=ON is now the default behavior.

These changes allow dependent libraries and frameworks to be able to call fork() et al prior to
initializing the underlying HSA and kernel runtimes since HCC intialization is deferred until first use.

### CODE REVIEW

**Review by scchan — COMMENTED** [review]
Would it be sufficient to set this to `false`?  https://github.com/RadeonOpenCompute/hcc/blob/clang_tot_upgrade/lib/mcwamp.cpp#L411

I had a PR to make that change but reverted it later due to a bug in HIP, which I think has been fixed.

**Comment by jeffdaily** [comment]
I would have simply set line 411 to false, but I wasn't sure about apps or libraries that might depend on the current default behavior of the immediate (i.e. non-lazy) init.  I reversed all the logic so that users could still access the old behavior by setting HCC_LAZYINIT=ON or HCC_LAZYINIT=0.  (I just pushed a fix for the latter.)

**Comment by jeffdaily** [comment]
Looks like I misunderstood the C++ static initialization order.   Setting HCC_LAZYINIT=ON was sufficient for full lazy initialization.  Should I revert the changes to the mcwamp_hsa.cpp file but leave the mcwamp.cpp changes?

**Comment by jeffdaily** [comment]
@scchan Please review again.   The changes are now only to the mcwamp.cpp source file to enable lazy init by default.

**Review by scchan — APPROVED** [review]
LGTM

**Comment by jeffdaily** [comment]
I am looking into the failed tests.  Such as 
```
terminate called after throwing an instance of 'Kalmar::runtime_exception'
  what():  HSA executable NOT built yet!
```

```
Failing Tests (5):
    HCC :: Unit/HC/accelerator_get_all_views_mt.cpp
    HCC :: Unit/HC/memcpy_symbol1.cpp
    HCC :: Unit/HC/memcpy_symbol2.cpp
    HCC :: Unit/HC/memcpy_symbol3.cpp
    HCC :: Unit/HC/memcpy_symbol4.cpp
```

**Comment by jeffdaily** [comment]
Down to 1 failure.
```
Failing Tests (1):
    HCC :: Unit/HC/accelerator_get_all_views_mt.cpp
```

**Comment by jeffdaily** [comment]
@scchan I expect all tests to pass now.  The final failing tested needed to be updated for the new lazy init default since it changes the number of accelerator views created.

File: lib/hsa/mcwamp_hsa.cpp [hunk 1]
```diff
@@ -2321,6 +2322,16 @@ class HSADevice final : public KalmarDevice
 
 private:
 
+    void load_executables_once() {
+        if (!executables_loaded) {
```
Comment (scchan) [inline lib/hsa/mcwamp_hsa.cpp]: I think this is racy and probably needs to be protected by something like `std::call_once` ?

**Review by scchan — CHANGES_REQUESTED** [review]

**Comment by jeffdaily** [comment]
I'm not sure how our CI is set up, but an unrelated commit triggered a rebuild of the PR-1138 jenkins build and it is now failing, even though it was passing on build 7 of this PR.

**Review by scchan — DISMISSED** [review]
LGTM in principle but now this PR has a merge conflict with the on-demand signal allocation patch you submitted earlier, could you take a look?

**Comment by jeffdaily** [comment]
@scchan I don't know that the failed test is due to the PR.  Should we have jenkins retest?  All tests were passing on my local system.

**Review by scchan — APPROVED** [review]
LGTM

**Pull request merged**